File: src/test/java/com/nickmcdowall/lsd/http/autoconfigure/LsdFeignAutoConfigurationTest.java
package com.nickmcdowall.lsd.http.autoconfigure;

import com.googlecode.yatspec.state.givenwhenthen.TestState;
import com.nickmcdowall.lsd.http.interceptor.LsdFeignLoggerInterceptor;
import feign.Logger;
import org.junit.jupiter.api.Test;
import org.springframework.boot.autoconfigure.AutoConfigurations;
import org.springframework.boot.test.context.assertj.AssertableApplicationContext;
import org.springframework.boot.test.context.runner.ApplicationContextRunner;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class LsdFeignAutoConfigurationTest {

    private final ApplicationContextRunner contextRunner = new ApplicationContextRunner()
            .withConfiguration(AutoConfigurations.of(
                    LsdFeignAutoConfiguration.class
            ));

    @Test
    public void noBeansAutoLoadedWhenRequiredBeansMissing() {
        contextRunner.withUserConfiguration(UserConfigWithoutRequiredBeans.class)
                .run((context) -> assertBeansNotLoaded(context));
    }

    @Test
    void loadsBeansWhenTestStateIsAvailable() {
        contextRunner.withUserConfiguration(UserConfigWithRequiredBeans.class).run((context) -> {
            assertThat(context).hasBean("defaultSourceNameMapping");
            assertThat(context).hasBean("defaultDestinationNameMapping");
            assertThat(context).hasBean("httpInteractionHandlers");
            assertThat(context).hasSingleBean(Logger.Level.class);
            assertThat(context).hasSingleBean(LsdFeignLoggerInterceptor.class);
        });
    }

    @Test
    void noBeansAutoLoadedWhenInterceptorsDisabledViaPropertyEvenIfBeansAvailable() {
        contextRunner.withUserConfiguration(UserConfigWithRequiredBeans.class)
                .withPropertyValues("yatspec.lsd.interceptors.autoconfig.enabled=false")
                .run((context) -> assertBeansNotLoaded(context));
    }

    @Test
    void doesntReplaceExistingLoggerBean() {
        contextRunner.withUserConfiguration(UserConfigWithExistingLoggerBean.class).run((context) -> {
            assertThat(context).getBean(Logger.Level.class)
                    .isEqualTo(new UserConfigWithExistingLoggerBean().feignLoggerLevel());
        });
    }

    private void assertBeansNotLoaded(AssertableApplicationContext context) {
        assertThat(context).doesNotHaveBean("defaultSourceNameMapping");
        assertThat(context).doesNotHaveBean("defaultDestinationNameMapping");
        assertThat(context).doesNotHaveBean("httpInteractionHandlers");
        assertThat(context).doesNotHaveBean(Logger.Level.class);
        assertThat(context).doesNotHaveBean(LsdFeignLoggerInterceptor.class);
    }

    @Configuration
    static class UserConfigWithoutRequiredBeans {
    }

    @Configuration
    static class UserConfigWithRequiredBeans {
        @Bean
        public TestState interactions() {
            return new TestState();
        }

        /*
         * To catch autoconfig beans of type List (the generic type is not taken into account so we need to use a name
         * or wrapper type for the collection
         */
        @Bean
        public List<Object> genericList() {
            return List.of();
        }
    }

    static class UserConfigWithExistingLoggerBean {
        @Bean
        public TestState interactions() {
            return new TestState();
        }

        @Bean
        public Logger.Level feignLoggerLevel() {
            return Logger.Level.FULL;
        }
    }
}

File: lsd-interceptors/src/test/java/io/lsdconsulting/interceptors/http/autoconfigure/LsdFeignAutoConfigurationTest.kt
package io.lsdconsulting.interceptors.http.autoconfigure

import feign.Logger
import io.lsdconsulting.interceptors.http.LsdFeignLoggerInterceptor
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.boot.autoconfigure.AutoConfigurations
import org.springframework.boot.test.context.assertj.AssertableApplicationContext
import org.springframework.boot.test.context.runner.ApplicationContextRunner
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

internal class LsdFeignAutoConfigurationTest {
    private val contextRunner = ApplicationContextRunner()
        .withConfiguration(
            AutoConfigurations.of(
                LsdFeignAutoConfiguration::class.java
            )
        )

    @Test
    fun loadsBeansWhenTestStateIsAvailable() {
        contextRunner.withUserConfiguration(UserConfigWithRequiredBeans::class.java)
            .run { context: AssertableApplicationContext? ->
                assertThat(context).hasBean("defaultSourceNameMapping")
                assertThat(context).hasBean("defaultDestinationNameMapping")
                assertThat(context).hasBean("httpInteractionHandlers")
                assertThat(context).hasSingleBean(
                    Logger.Level::class.java
                )
                assertThat(context).hasSingleBean(
                    LsdFeignLoggerInterceptor::class.java
                )
            }
    }

    @Test
    fun noBeansAutoLoadedWhenInterceptorsDisabledViaPropertyEvenIfBeansAvailable() {
        contextRunner.withUserConfiguration(UserConfigWithRequiredBeans::class.java)
            .withPropertyValues("lsd.interceptors.autoconfig.enabled=false")
            .run { context: AssertableApplicationContext -> assertBeansNotLoaded(context) }
    }

    @Test
    fun doesntReplaceExistingLoggerBean() {
        contextRunner.withUserConfiguration(UserConfigWithExistingLoggerBean::class.java)
            .run { context: AssertableApplicationContext? ->
                assertThat(context).getBean(Logger.Level::class.java)
                    .isEqualTo(UserConfigWithExistingLoggerBean().feignLoggerLevel())
            }
    }

    private fun assertBeansNotLoaded(context: AssertableApplicationContext) {
        assertThat(context).doesNotHaveBean("defaultSourceNameMapping")
        assertThat(context).doesNotHaveBean("defaultDestinationNameMapping")
        assertThat(context).doesNotHaveBean("httpInteractionHandlers")
        assertThat(context).doesNotHaveBean(Logger.Level::class.java)
        assertThat(context).doesNotHaveBean(LsdFeignLoggerInterceptor::class.java)
    }

    @Configuration
    internal open class UserConfigWithoutRequiredBeans

    @Configuration
    internal open class UserConfigWithRequiredBeans {
        /*
         * To catch autoconfig beans of type List (the generic type is not taken into account, so we need to use a name
         * or wrapper type for the collection
         */
        @Bean
        open fun genericList(): List<Any> {
            return listOf()
        }
    }

    internal class UserConfigWithExistingLoggerBean {
        @Bean
        fun feignLoggerLevel(): Logger.Level {
            return Logger.Level.FULL
        }
    }
}
